### PR TITLE
[evdev 2/3] Introduce evdev - generic input event interface

### PR DESCRIPTION
**Pull request by xThaid**
The part of the #1089 PR containing only the implementation of evdev. Should be merged after #1093.

### CODE REVIEW

**Comment by hadarai** [comment]
Please write all comments in [Doxygen convection](https://www.doxygen.nl/manual/docblocks.html)
(the one with `\brief`, `\summary` etc.)

**Review by hadarai — CHANGES_REQUESTED** [review]
Please write all comments in [Doxygen convection](https://www.doxygen.nl/manual/docblocks.html)
(the one with `\brief`, `\summary` etc.)

File: sys/drv/atkbdc.c [hunk 1]
```diff
@@ -61,28 +66,27 @@ static void write_data(resource_t *regs, uint8_t byte) {
   bus_write_1(regs, KBD_DATA_PORT, byte);
 }
 
-static int scancode_read(vnode_t *v, uio_t *uio, int ioflag) {
-  atkbdc_state_t *atkbdc = devfs_node_data(v);
-  int error;
+static void atkbdc_thread(void *arg) {
```
Comment (hadarai) [inline sys/drv/atkbdc.c:69]: I'm afraid that function with that name really needs a comment of explanation

File: sys/drv/atkbdc.c [hunk 2]
```diff
@@ -145,6 +149,20 @@ static int atkbdc_probe(device_t *dev) {
   return 1;
 }
 
+static void evdev_init(atkbdc_state_t *atkbdc) {
```
Comment (hadarai) [inline sys/drv/atkbdc.c:152]: That init does so much that I'd like to see a brief of what is initialised

**Review by hadarai — CHANGES_REQUESTED** [review]

**Comment by hadarai** [comment]
> Please write all comments in [Doxygen convection](https://www.doxygen.nl/manual/docblocks.html)
> (the one with `\brief`, `\summary` etc.)

According to decisions from latest meeting that request is outdated.

File: include/dev/evdev.h [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+ * Convert a USB keyboard scancode to a evdev-compatible keycode.
+ */
+uint16_t evdev_hid2key(int scancode);
+
+/*
+ * Convert a AT keyboard scancode to a evdev-compatible keycode. The function is
+ * stateful because of AT extended scan codes - some keys are encoded in two
+ * numbers.
+ */
+uint16_t evdev_scancode2key(int *state, int scancode);
```
[the remark this answers is not in the document]
Comment (cahirwpz) [reply include/dev/evdev.h:91]: `state` should be renamed to `statep` where `p` stands for `pointer`

File: include/dev/evdev.h [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+/*
+ * Set a flag describing supported features by the evdev device.
+ */
+void evdev_set_flag(evdev_dev_t *evdev, uint16_t flag);
+
+/*
+ * Finalize initialization and register an evdev device. Should be called after
+ * all properties of the evdev device are set.
+ */
+int evdev_register(evdev_dev_t *evdev);
```
[the remark this answers is not in the document]
Comment (xThaid) [reply include/dev/evdev.h:71]: Ok, added a comment.

File: include/dev/evdev.h [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+/*
+ * Push a new event to an evdev.
+ */
+int evdev_push_event(evdev_dev_t *evdev, uint16_t type, uint16_t code,
```
[the remark this answers is not in the document]
Comment (xThaid) [reply include/dev/evdev.h:76]: Done

File: include/dev/evdev.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (xThaid) [reply include/dev/evdev.h:91]: Done

File: sys/drv/evdev_kbd.c [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+	NONE,		NONE,		NONE,		NONE,
+	NONE,		NONE,		NONE,		NONE,
+};
+/* clang-format on */
+
+uint16_t evdev_hid2key(int scancode) {
+  return evdev_usb_scancodes[scancode];
+}
+
+uint16_t evdev_scancode2key(int *state, int scancode) {
```
[the remark this answers is not in the document]
Comment (xThaid) [reply sys/drv/evdev_kbd.c:160]: It is already described in the header file.

File: sys/drv/evdev_kbd.c [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+      if ((scancode & 0x7f) != 0x45)
+        return NONE;
+      keycode = KEY_PAUSE;
+      break;
+  }
+
+  return keycode;
+}
+
+void evdev_support_all_known_keys(evdev_dev_t *evdev) {
```
[the remark this answers is not in the document]
Comment (xThaid) [reply sys/drv/evdev_kbd.c:204]: It is already described in the header file.

File: sys/drv/atkbdc.c (hunk 1 above)
Comment (xThaid) [reply sys/drv/atkbdc.c:69]: Done

File: sys/drv/atkbdc.c (hunk 2 above)
Comment (xThaid) [reply sys/drv/atkbdc.c:157]: Sure, I've added brief comments. Hope it helps.

**Review by franciscozdo — CHANGES_REQUESTED** [review]

File: sys/drv/evdev.c [hunk 8]
```diff
@@ -0,0 +1,644 @@
...
+  int ret, unit = 0;
+
+  /*
+   * Find the first free number for that device. Not the best solution, but will
+   * do for now
+   */
+  do {
+    snprintf(buf, sizeof(buf), "event%d", unit);
+    ret = devfs_makedev(evdev_input_dir, buf, &evdev_vnodeops, evdev, NULL);
+  } while (ret == EEXIST);
```
[the remark this answers is not in the document]
Comment (xThaid) [reply sys/drv/evdev.c:429]: Yeah, of course. Thanks.

File: sys/drv/evdev.c [hunk 9]
```diff
@@ -0,0 +1,644 @@
...
+ * general informations about the device and its state. A name, device ID and
+ * supported features flags are only used by user-space.
+ *
+ * `evdev_client_t` is created whenever the evdev file is opened. Its jobs is to
+ * handle an interaction with the user-space. Every client has its private event
+ * buffer. Every time an input event is issued, it firstly goes through
+ *`evdev_dev_t` and then it is propagated to clients. This allows us to have
+ * multiple, non-interfering readers of a single evdev device. Hence, every
+ *input device has exactly one `evdev_dev_t`, but possibly can have many
+ *`evdev_client_t` structures linked to it.
```
[the remark this answers is not in the document]
Comment (xThaid) [reply sys/drv/evdev.c:54]: +1

File: sys/drv/evdev.c [hunk 10]
```diff
@@ -0,0 +1,644 @@
...
+      (input_event_t){.type = EV_SYN, .code = SYN_DROPPED, .value = 0};
+    client->ec_buffer_head = head;
+    client->ec_buffer_ready = head;
+  }
+
+  client->ec_buffer[tail].type = type;
+  client->ec_buffer[tail].code = code;
+  client->ec_buffer[tail].value = value;
+
+  tail = (tail == count - 1) ? 0 : (tail + 1);
```
[the remark this answers is not in the document]
Comment (xThaid) [reply sys/drv/evdev.c:222]: Yeah, there is. Cahir asked me to get rid of the modulo operation, so I did it.

File: sys/drv/evdev_kbd.c [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+       * Ctrl-pause/break produces:
+       * E0-46 E0-C6 (See above.)
+       */
+      *statep = 0;
+      if ((scancode & 0x7f) == 0x1D)
+        *statep = scancode;
+      return NONE;
+      /* NOT REACHED */
+    case 0x1D: /* pause / break */
+    case 0x9D:
```
[the remark this answers is not in the document]
Comment (xThaid) [reply sys/drv/evdev_kbd.c:191]: To be honest, I am not sure if these macros would be useful. As far as I know they don't have a special names, so macros like `#define E0_PREFIX 0xE0` wouldn't be helpful. Another argument is that the macros would be used only in one place, so again I don't see real benefits.

[the remark this answers is not in the document]
Comment (franciscozdo) [reply sys/drv/evdev_kbd.c:191]: Okay. It's good then.

File: sys/drv/evdev.c [hunk 12]
```diff
@@ -0,0 +1,644 @@
...
+    head = ready = tail = 0;
+    client->ec_buffer[head] =
+      (input_event_t){.type = EV_SYN, .code = SYN_DROPPED, .value = 0};
+    client->ec_buffer_head = head;
+    client->ec_buffer_ready = head;
+  }
+
+  client->ec_buffer[tail].type = type;
+  client->ec_buffer[tail].code = code;
+  client->ec_buffer[tail].value = value;
```
[the remark this answers is not in the document]
Comment (xThaid) [reply sys/drv/evdev.c:220]: You're right, thanks.

File: sys/drv/evdev.c [hunk 13]
```diff
@@ -0,0 +1,644 @@
...
+  evdev_dev_t *evdev = (evdev_dev_t *)arg;
+
+  evdev_send_event(evdev, EV_KEY, evdev->ev_rep_key, KEY_EVENT_REPEAT);
+  evdev_send_event(evdev, EV_SYN, SYN_REPORT, 1);
+
+  if (evdev->ev_rep[REP_PERIOD])
+    callout_reschedule(&evdev->ev_rep_callout,
+                       evdev->ev_rep[REP_PERIOD] * CLK_TCK / 1000);
+  else
+    evdev->ev_rep_key = KEY_RESERVED;
```
[the remark this answers is not in the document]
Comment (xThaid) [reply sys/drv/evdev.c:530]: Yup, I should. I forgot it will be called from another thread.

**Review by franciscozdo — APPROVED** [review]

**Pull request merged**